core/types/subtyping.cc

```diff
@@ -1211,25 +1220,47 @@ bool isSubTypeUnderConstraintSingle(const GlobalState &gs, TypeConstraint &const
 
                 ENFORCE(i < a1->klass.data(gs)->typeMembers().size());
 
+                auto &a = a1->targs[i];
+                auto &b = a2->targs[j];
+                bool doesMemberMatch = true;
+                string variance;
+                string joiningText;
+                auto subCollector = errorDetailsCollector.newCollector();
                 if (idxTypeMember.data(gs)->flags.isCovariant) {
-                    result = Types::isSubTypeUnderConstraint(gs, constr, a1->targs[i], a2->targs[j], mode);
+                    variance = "covariant";
+                    joiningText = "a subtype of";
+                    doesMemberMatch = Types::isSubTypeUnderConstraint(gs, constr, a, b, mode, subCollector);
                 } else if (idxTypeMember.data(gs)->flags.isInvariant) {
-                    auto &a = a1->targs[i];
-                    auto &b = a2->targs[j];
+                    variance = "invariant";
+                    joiningText = "equivalent to";
                     if (mode == UntypedMode::AlwaysCompatible) {
-                        result = Types::equivUnderConstraint(gs, constr, a, b);
+                        doesMemberMatch = Types::equivUnderConstraint(gs, constr, a, b, subCollector);
                     } else {
                         // At the time of writing, we never set mode == UntypedMode::AlwaysIncompatible
                         // except when `constr` is EmptyFrozenConstraint, so there's no observable
                         // difference whether we use equivNoUntyped or equivNoUntypedUnderConstraint here.
                         // May as well do it for symmetry though.
-                        result = Types::equivNoUntypedUnderConstraint(gs, constr, a, b);
+                        doesMemberMatch = Types::equivNoUntypedUnderConstraint(gs, constr, a, b, subCollector);
                     }
                 } else if (idxTypeMember.data(gs)->flags.isContravariant) {
-                    result = Types::isSubTypeUnderConstraint(gs, constr, a2->targs[j], a1->targs[i], mode);
+                    variance = "contravariant";
+                    joiningText = "a supertype of";
+                    doesMemberMatch = Types::isSubTypeUnderConstraint(gs, constr, b, a, mode, subCollector);
                 }
-                if (!result) {
-                    break;
+                if (!doesMemberMatch) {
+                    result = false;
+                    if constexpr (std::is_same<T, ErrorDetailsCollector>::value) {
+                        // TODO(neil): once we pass in whether this a covariant or contravariant call,
+                        // we should reword it in the "expected ... but got ..." form
+                        // TODO(neil): if the type member is for a Proc we should have special wording,
+                        // ie. Proc return value instead of Proc1::Return and Proc arg 1 instead of Proc1::Arg1
```

Comment:
I don't think we necessarily have to do this for Hash and Array—people recognize that `Hash` is a generic class with type arguments, because users are forced to provide these every time they make a typed `Hash`.

The difference with `T.proc` is that people think that `T.proc` is something "built into Sorbet," and not just a generic class under the hood.

We _might_ want to silence the error about `Hash::Elem` specifically, because it's a bit weird, but I could go either way on that. Based on the way it's created, if either of `K` or `V` have a problem, then so will `Elem`, so it'll always be duplicated and yeah probably makes sense to be hidden.